olm/operator-bundle.md

```diff
@@ -0,0 +1,199 @@
+
+For an operator registry bundle, the source directory structure will have similar layout as shown below.
+```
+$ tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+└── annotations.yaml
+```
+
+`Dockerfile` for operator bundle
+```
+FROM scratch
+
+# We are pushing an operator-registry bundle that
+# has both metadata and manifests.
+LABEL operators.coreos.com.bundle.resources=manifests+metadata
+LABEL operators.coreos.com.bundle.mediatype=bundle
+
+ADD test/0.1.0 /manifests
+ADD test/annotations.yaml /metadata/annotations.yaml
+```
+
+Below is the directory layout of the operator bundle insdie the image.
+```bash
+$ tree
+/
+├── manifests
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+└── metadata
+    └── annotations.yaml
+```
+
+###### UX:
+Build, Push and Pull an operator bundle.
+```
+docker build -f Dockerfile -t quay.io/test/test-operator:v1 .
+docker push quay.io/test/test-operator:v1
+docker pull quay.io/test/test-operator:v1
+```
+
+A tool can inspect an operator bundle to determine the bundle type and its media.
+```bash
+docker image inspect quay.io/test/test-operator:v1 | \
+jq '.[0].Config.Labels["operators.coreos.com.bundle.resources"]'
+
+"manifests+metadata"
+```
+
+### Verify, Run and Test 
+
+#### Generate Scaffolding
+As an operator author I want to generate the scaffolding necessary to create an operator bundle. We provide the operator author to automatically generate the scaffolding.
+```bash
+$ tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+
+$ cd test
+
+# the following command generates the necessary scaffolding.
+$ operator-framework bundle init --type=operator-registry --bundle-folder=0.1.0
+
+# output:
+#  - test/Dockerfile
+#  - test/annotations.yaml
+```
+
+Once the scaffolding is done, the user can do a `docker build` to create an operator bundle.
+
+#### Validate an Operator Bundle
+As an operator author I want to validate an operator bundle so that I can ensure it runs as expected on a cluster.
+```bash
+$ operator-framework bundle validate --image=quay.io/test/test-operator:v1
+```
+
+The validate command will do the following:
+* Make sure the image `label` and `annotations.yaml` are appropriately configured.
+* Validate the operator registry bundle.
+
+#### Run the Operator from the Bundle 
+As an operator author I want to run my operator directly from the operator bundle image.
+```bash
+cat <<EOF | kubectl apply -f -
+apiVersion: operators.coreos.com/v2alpha1
+kind: Operator
```

Comment:
What component would watch for this Operator resource and translate this into a running operator.

Reply:
This example shows how a user can run the operator using `olm`. This is being covered here - https://github.com/openshift/enhancements/pull/28.

Reply:
@tkashem I think it would also be useful to write out manual application by pulling the bundle image locally, unpacking it, and applying to the cluster.